posts/web/2017-01-06-dep-yandex-direct.md

```diff
@@ -0,0 +1,198 @@
+---
+author: Роман Киреев
+title:  "Зависимые типы на примере взаимодействия с Яндекс.Директ API"
+tags:   зависимые типы, семейства типов, web, servant
+description: На примере Яндекс.Директ API в статье показывается, как типизировать такой запрос к API, тип результата которого зависит от содержимого запроса.
+---
+
+Здравствуйте.
+
+Зависимые типы набирают популярность и в будущем [будут добавлены](https://typesandkinds.wordpress.com/2016/07/24/dependent-types-in-haskell-progress-report) в GHC. Однако, зависимые типы ассоциируются с доказательством теорем и формализацией различных систем, так что их польза может быть неочевидной для рядового программиста. В этой статье мы рассмотрим зависимые типы на примере простой программистской задачи -- типобезопасного взаимодействия с Яндекс.Директ API v5.
+
+Популяризации зависимых типов для программирования немало способствует библиотека Servant, используемая как для предоставления типизированного API (servant-server), так и для написания запросов к уже существующему (servant-client). Мы будем использовать эту библиотеку, но я не стану объяснять принцип ее работы, поскольку такой доклад уже [есть](https://ruhaskell.org/posts/talks/2016/04/05/better-services-with-servant.html). Однако, встроенными в Servant зависимыми типами дело не ограничится, и мы добавим еще своих.
```

Comment:
@VyacheslavHashov, [Dependently typed servers in Servant](https://www.well-typed.com/blog/2015/12/dependently-typed-servers/).